core/server/proxy/src/test/java/alluxio/StreamCacheTest.java

```diff
@@ -47,70 +48,127 @@ public void concurrentOperations() throws Exception {
-    final AtomicInteger isId = new AtomicInteger(0);
-    Runnable putIs = new Runnable() {
+    final AtomicIntegerArray isIDs = new AtomicIntegerArray(numThreads);
+    final AtomicInteger isIDsIndex = new AtomicInteger();
+    final AtomicIntegerArray osIDs = new AtomicIntegerArray(numThreads);
+    final AtomicInteger osIDsIndex = new AtomicInteger();
+    Runnable put = new Runnable() {
       @Override
       public void run() {
-        isId.set(streamCache.put(is));
+        int id = streamCache.put(is);
+        int index = isIDsIndex.getAndIncrement();
+        isIDs.set(index, id);
+        id = streamCache.put(os);
+        index = osIDsIndex.getAndIncrement();
+        osIDs.set(index, id);
       }
     };
-    final AtomicInteger osId = new AtomicInteger(0);
-    Runnable putOs = new Runnable() {
-      @Override
-      public void run() {
-        osId.set(streamCache.put(os));
-      }
-    };
-    Thread putIsThread = new Thread(putIs);
-    Thread putOsThread = new Thread(putOs);
-    putIsThread.start();
-    putOsThread.start();
-    putIsThread.join();
-    putOsThread.join();
-    Assert.assertSame(is, streamCache.getInStream(isId.get()));
-    Assert.assertSame(os, streamCache.getOutStream(osId.get()));
+    Thread[] threads = new Thread[numThreads];
+    for (int i = 0; i < numThreads; i++) {
+      Thread thread = new Thread(put);
+      threads[i] = thread;
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    Assert.assertEquals(numThreads, isIDsIndex.get());
+    Assert.assertEquals(numThreads, osIDsIndex.get());
+    for (int i = 0; i < numThreads; i++) {
+      Assert.assertSame(is, streamCache.getInStream(isIDs.get(i)));
+      Assert.assertSame(os, streamCache.getOutStream(osIDs.get(i)));
+    }
+    Assert.assertEquals(2 * numThreads, streamCache.size());
 
     // Concurrent get.
-    Runnable getIs = new Runnable() {
+    class Get implements Runnable {
+      private int mThreadIndex;
+
+      Get(int threadIndex) {
+        mThreadIndex = threadIndex;
+      }
+
       @Override
       public void run() {
-        Assert.assertSame(is, streamCache.getInStream(isId.get()));
+        Assert.assertSame(is, streamCache.getInStream(isIDs.get(mThreadIndex)));
+        Assert.assertSame(os, streamCache.getOutStream(osIDs.get(mThreadIndex)));
       }
-    };
-    Runnable getOs = new Runnable() {
+    }
+    for (int i = 0; i < numThreads; i++) {
+      threads[i] = new Thread(new Get(i));
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    for (int i = 0; i < numThreads; i++) {
+      Assert.assertSame(is, streamCache.getInStream(isIDs.get(i)));
+      Assert.assertSame(os, streamCache.getOutStream(osIDs.get(i)));
+    }
+    Assert.assertEquals(2 * numThreads, streamCache.size());
+
+    // Concurrent get, put, and invalidate.
+    class GetPutInvalidate implements Runnable {
```

Comment:
ditto, have separate operations for input and output stream to better test the interleaving of such operations